app/src/main/kotlin/com/wire/android/ui/calling/SharedCallingViewModel.kt

```diff
@@ -279,8 +310,48 @@ class SharedCallingViewModel @AssistedInject constructor(
         }
     }
 
+    private fun observeInCallReactions() {
+        observeInCallReactionsUseCase(conversationId).onEach { message ->
+
+            val sender = participantsState.senderName(message.senderUserId)?.let { name ->
+                ReactionSender.Other(name)
+            } ?: ReactionSender.Unknown
+
+            message.emojis.forEach { emoji ->
+                _inCallReactions.send(InCallReaction(emoji, sender))
+            }
+
+            if (message.emojis.isNotEmpty()) {
+                recentReactions.put(message.senderUserId, message.emojis.last())
+            }
+        }.launchIn(viewModelScope)
+    }
+
+    fun onReactionClick(emoji: String) {
+        viewModelScope.launch {
+            sendInCallReactionUseCase(conversationId, emoji).onSuccess {
+                _inCallReactions.send(InCallReaction(emoji, ReactionSender.You))
+            }
+        }
+    }
+
+    private fun recentInCallReactionMap(): MutableMap<UserId, String> =
+        ExpiringMap<UserId, String>(
+            scope = viewModelScope,
+            expiration = InCallReactions.recentReactionShowDurationMs,
+            delegate = mutableStateMapOf<UserId, String>()
+        )
+
     @AssistedFactory
     interface Factory {
         fun create(conversationId: ConversationId): SharedCallingViewModel
     }
 }
+
+private fun List<UICallParticipant>.senderName(userId: QualifiedID) = firstOrNull { it.id.value == userId.value }?.name
+
+private fun <T> Flow<T>.withDelayAfterFirst(timeMillis: Long): Flow<T> = withIndex()
```

Comment:
can you move it to `com.wire.android.util.extension.Flow.kt` so It might be re-used somewhere else